### PR TITLE
Upgrade jib-build-plan to 0.3.1

### PR DESCRIPTION
**Pull request by chanseokoh**
jib-build-plan 0.3.1 is now onilne.

### CODE REVIEW

**Comment by chanseokoh** [comment]
Ugh, right. This is a breaking change.

File: jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java [hunk 1]
```diff
@@ -222,8 +222,7 @@ public void testToContainerBuildPlan_default() throws InvalidImageReferenceExcep
 
     ContainerBuildPlan buildPlan = containerBuilder.toContainerBuildPlan();
     Assert.assertEquals("base/image", buildPlan.getBaseImage());
-    Assert.assertEquals("amd64", buildPlan.getArchitectureHint());
-    Assert.assertEquals("linux", buildPlan.getOsHint());
+    ;
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java:225]: Should remove the line container the stray `;`. However, I think we should rather test `getPlatforms()`.

File: jib-core/src/main/java/com/google/cloud/tools/jib/api/JibContainerBuilder.java [hunk 2]
```diff
@@ -610,8 +609,6 @@ public JibContainerBuilder applyContainerBuildPlan(ContainerBuildPlan buildPlan)
       throws InvalidImageReferenceException {
     containerBuildPlanBuilder
         .setBaseImage(buildPlan.getBaseImage())
-        .setArchitectureHint(buildPlan.getArchitectureHint())
-        .setOsHint(buildPlan.getOsHint())
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/api/JibContainerBuilder.java:614]: I think this should call `setPlatforms(buildPlan.getPlatforms())`.

File: jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java [hunk 3]
```diff
@@ -257,8 +256,6 @@ public void testToContainerBuildPlan() throws InvalidImageReferenceException, IO
 
     ContainerBuildPlan buildPlan = containerBuilder.toContainerBuildPlan();
     Assert.assertEquals("base/image", buildPlan.getBaseImage());
-    Assert.assertEquals("amd64", buildPlan.getArchitectureHint());
-    Assert.assertEquals("linux", buildPlan.getOsHint());
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java:261]: ditto. We should test `getPlatforms()`.

File: jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java [hunk 4]
```diff
@@ -297,8 +294,6 @@ public void setApplyContainerBuildPlan()
     ContainerBuildPlan buildPlan =
         ContainerBuildPlan.builder()
             .setBaseImage("some/base")
-            .setArchitectureHint("arch")
-            .setOsHint("os")
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java:301]: Perhaps `setPlatforms(Collections.singleton(new Platform("arch", "os")))`.

File: jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java [hunk 5]
```diff
@@ -349,9 +344,5 @@ public void setApplyContainerBuildPlan()
         AbsoluteUnixPath.get("/workspace"), containerConfiguration.getWorkingDirectory());
     Assert.assertEquals(Arrays.asList("foo", "entrypoint"), containerConfiguration.getEntrypoint());
     Assert.assertEquals(Arrays.asList("bar", "cmd"), containerConfiguration.getProgramArguments());
-
-    ContainerBuildPlan convertedPlan = containerBuilder.toContainerBuildPlan();
-    Assert.assertEquals("arch", convertedPlan.getArchitectureHint());
-    Assert.assertEquals("os", convertedPlan.getOsHint());
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java:355]: We should keep the test (with `getPlatforms()`).

**Comment by louismurerwa** [comment]
@chanseokoh I had already integrated ` getPlatforms() ` in this PR  https://github.com/GoogleContainerTools/jib/pull/2589

File: jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java (hunk 1 above)
Comment (louismurerwa) [reply jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java]: Done

File: jib-core/src/main/java/com/google/cloud/tools/jib/api/JibContainerBuilder.java (hunk 2 above)
Comment (louismurerwa) [reply jib-core/src/main/java/com/google/cloud/tools/jib/api/JibContainerBuilder.java:614]: I had done it here  https://github.com/GoogleContainerTools/jib/blob/0ab161a7787210925a2bb3af4d43884bfdeadcdc/jib-core/src/main/java/com/google/cloud/tools/jib/api/JibContainerBuilder.java

**Comment by chanseokoh** [comment]
Yeah, I am aware of that, but I think we shouldn't merge a partially working code into HEAD. The existing test and code should be preserved during migration. Each PR does what it really intends to do, which will make the process easy when we need to revert each step. This PR is for upgrading the build plan to 0.3.0, while #2589 is to further modify the jib-core API.

**Comment by louismurerwa** [comment]
Ohh I see ,let me make the edits and push them up

File: jib-core/src/main/java/com/google/cloud/tools/jib/api/JibContainerBuilder.java [hunk 6]
```diff
@@ -610,8 +633,8 @@ public JibContainerBuilder applyContainerBuildPlan(ContainerBuildPlan buildPlan)
...
+        .setPlatforms(buildPlan.getPlatforms())
```
Comment (louismurerwa) [inline jib-core/src/main/java/com/google/cloud/tools/jib/api/JibContainerBuilder.java:636]: Remove one of the `setPlatforms`

File: jib-core/src/main/java/com/google/cloud/tools/jib/api/JibContainerBuilder.java [hunk 7]
```diff
@@ -489,6 +489,29 @@ public JibContainerBuilder setCreationTime(Instant creationTime) {
...
+   * Sets the platforms configurations for the container.
+   *
+   * @param platforms the list of platform objects
+   * @return this
+   */
+  public JibContainerBuilder setPlatforms(Set<Platform> platforms) {
+    containerBuildPlanBuilder.setPlatforms(platforms);
+    return this;
+  }
+
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/api/JibContainerBuilder.java:514]: Do not add new methods to `JibContainerBuidler` in this PR. You'll do it in #2589.

Comment (louismurerwa) [reply jib-core/src/main/java/com/google/cloud/tools/jib/api/JibContainerBuilder.java]: Done

File: jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java [hunk 8]
```diff
@@ -351,7 +348,7 @@ public void setApplyContainerBuildPlan()
     Assert.assertEquals(Arrays.asList("bar", "cmd"), containerConfiguration.getProgramArguments());
 
     ContainerBuildPlan convertedPlan = containerBuilder.toContainerBuildPlan();
-    Assert.assertEquals("arch", convertedPlan.getArchitectureHint());
-    Assert.assertEquals("os", convertedPlan.getOsHint());
+    Assert.assertEquals(
+        ImmutableSet.of(new Platform("testOS", "testArchitecture")), convertedPlan.getPlatforms());
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java:352]:
```suggestion
        ImmutableSet.of(new Platform("testArchitecture", "testOs")), convertedPlan.getPlatforms());
```

File: jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java [hunk 9]
```diff
@@ -297,8 +295,7 @@ public void setApplyContainerBuildPlan()
     ContainerBuildPlan buildPlan =
         ContainerBuildPlan.builder()
             .setBaseImage("some/base")
-            .setArchitectureHint("arch")
-            .setOsHint("os")
+            .setPlatforms(ImmutableSet.of(new Platform("testOS", "testArchitecture")))
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java:298]: argument order

File: jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java [hunk 10]
```diff
@@ -257,8 +256,7 @@ public void testToContainerBuildPlan() throws InvalidImageReferenceException, IO
 
     ContainerBuildPlan buildPlan = containerBuilder.toContainerBuildPlan();
     Assert.assertEquals("base/image", buildPlan.getBaseImage());
-    Assert.assertEquals("amd64", buildPlan.getArchitectureHint());
-    Assert.assertEquals("linux", buildPlan.getOsHint());
+    Assert.assertEquals(ImmutableSet.of(new Platform("linux", "amd64")), buildPlan.getPlatforms());
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java:259]: argument order

File: jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java [hunk 11]
```diff
@@ -222,8 +223,6 @@ public void testToContainerBuildPlan_default() throws InvalidImageReferenceExcep
 
     ContainerBuildPlan buildPlan = containerBuilder.toContainerBuildPlan();
     Assert.assertEquals("base/image", buildPlan.getBaseImage());
-    Assert.assertEquals("amd64", buildPlan.getArchitectureHint());
-    Assert.assertEquals("linux", buildPlan.getOsHint());
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java:226]: Restore the existing tests.

File: jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java (hunk 8 above)
Comment (louismurerwa) [reply jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java:352]: Did we already update the platform api to change the argument order? The one we deployed.

Comment (chanseokoh) [reply jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java:352]: Yeah. If you look into the jib-build-plan 3.0 library, you'll see the actual code.

image

Comment (louismurerwa) [reply jib-core/src/test/java/com/google/cloud/tools/jib/api/JibContainerBuilderTest.java]: Nice one.

**Comment by chanseokoh** [comment]
Tests are failing. Please fix them. I don't think the values `testArchitecture` and `testOs` is set anywhere.

**Comment by louismurerwa** [comment]
I have been going through the files to see why the tests are failing .The reason is that in our api the default platform is ` linux/amd64 ` isntead of `amd64\linux` .We might need to tweak our api a little bit.

**Comment by chanseokoh** [comment]
Really? That doesn't make sense to me. We need to double-check.

**Comment by louismurerwa** [comment]
Here : https://github.com/GoogleContainerTools/jib/blob/b3f130648f8c892edcc30bb83d115e14e6613bee/jib-build-plan/src/main/java/com/google/cloud/tools/jib/api/buildplan/ContainerBuildPlan.java#L44

**Comment by chanseokoh** [comment]
What a disaster. Let's fix that and release jib-build-plan 0.3.1. Could your file a PR to fix that?

**Comment by louismurerwa** [comment]
Yes I will fix it now

File: jib-core/CHANGELOG.md [hunk 12]
```diff
@@ -9,6 +9,8 @@ All notable changes to this project will be documented in this file.
 
 ### Changed
 
+- Upgraded jib-build-plan to 0.3.0. ([#2594](https://github.com/GoogleContainerTools/jib/pull/2594))
```
Comment (chanseokoh) [inline jib-core/CHANGELOG.md:12]: 0.3.1

File: jib-maven-plugin-extension-api/CHANGELOG.md [hunk 13]
```diff
@@ -7,5 +7,7 @@ All notable changes to this project will be documented in this file.
 
 ### Changed
 
+- Upgraded jib-build-plan to 0.3.0. ([#2594](https://github.com/GoogleContainerTools/jib/pull/2594))
```
Comment (chanseokoh) [inline jib-maven-plugin-extension-api/CHANGELOG.md:10]: ditto

File: jib-gradle-plugin-extension-api/CHANGELOG.md [hunk 14]
```diff
@@ -7,5 +7,7 @@ All notable changes to this project will be documented in this file.
 
 ### Changed
 
+- Upgraded jib-build-plan to 0.3.0. ([#2594](https://github.com/GoogleContainerTools/jib/pull/2594))
```
Comment (chanseokoh) [inline jib-gradle-plugin-extension-api/CHANGELOG.md:10]: ditto

File: jib-core/CHANGELOG.md (hunk 12 above)
Comment (louismurerwa) [reply jib-core/CHANGELOG.md:12]: Is this supposed to be a `CHANGELOG` for 0.3.0 though?

Comment (chanseokoh) [reply jib-core/CHANGELOG.md:12]: Good question. Let me explain. Note that the top CHANGELOG entry is `## [unreleased]` and the second entry is `## 0.15.0`. That is, the latest jib-core version we released and is available on the Internet (on the repository called Maven Central) is 0.15.0. Now, until we make a new jib-core release (which could be 0.16.0, for example), we write under `## [unreleased]` all important and publicly visible changes that the user needs/wants to know about that future release. The current jib-core 0.15.0 is using jib-build-plan 0.2.0, while the next jib-core release will use jib-build-plan 0.3.1. This is exactly what should be documented and also the truth.

**Comment by chanseokoh** [comment]
And don't forget to change the PR title: 0.3.0 --> 0.3.1

File: jib-core/CHANGELOG.md (hunk 12 above)
Comment (louismurerwa) [reply jib-core/CHANGELOG.md:12]: Thanks for the explanation @chanseokoh  ,now i understand how the release cycle works better.

**Comment by chanseokoh** [comment]
LGTM, approved

**Pull request merged**